### PR TITLE
tokenization_marian.py: use current_spm for decoding

### LINKED ISSUE #10294: Marian input decoding bug
# 🐛 Bug

## Information

Model I am using (Bert, XLNet ...): Marian

Language I am using the model on (English, Chinese ...): English, German

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

I'm examining Marian models to translate text. I noticed `convert_tokens_to_string` method uses `spm_target` which can be problematic if we want to decode source text.

Here is my script:

```
from transformers import MarianTokenizer, MarianModel

tokenizer = MarianTokenizer.from_pretrained('Helsinki-NLP/opus-mt-en-de')
model = MarianModel.from_pretrained('Helsinki-NLP/opus-mt-en-de')

input_text = "I was eating lunch when he saw me"
target_text = "Ich aß gerade zu Mittag, als er mich sah"

input_tokenized = tokenizer(input_text, return_tensors='pt')
with tokenizer.as_target_tokenizer():
    target_tokenized = tokenizer(target_text, return_tensors='pt')

print(tokenizer.decode(input_tokenized.data['input_ids'][0]))
with tokenizer.as_target_tokenizer():
    print(tokenizer.decode(target_tokenized.data['input_ids'][0]))

```

stdout:

```
I was▁eating▁lunch▁when he▁saw me
Ich aß gerade zu Mittag, als er mich sah

```

As you can see the input text is not decoded correctly since `spm_target` is used. A potential fix is to use `current_spm` and let `as_target_tokenizer` context manager decide which spm should be used (similar to text encoding):

```
def convert_tokens_to_string(self, tokens: List[str]) -> str:
    return self.current_spm.DecodePieces(tokens)
```

I can PR the fix if needed.

## Environment info

     
- `transformers` version: master branch (f6e53e3c2bafb37c861db71a4b28c304403af92b)
- Python version: 3.7.4
- PyTorch version (GPU?): 1.7.1 (False)

### PR DESCRIPTION
**Pull request by Mehrad0711**
# What does this PR do?

Fixes #10294 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Review by patil-suraj — COMMENTED** [review]
Hey @Mehrad0711, thanks a lot for the PR!

`current_spm` is always `spm_source` except inside the `as_target_tokenizer`. Now with this change, if we want to decode the targets, we'll need to do that inside the context manager, so we should also update all the examples in the docs and the `run_seq_to_seq.py` script to call `decode` or `batch_decode` inside the context manager to correctly decode the targets.

**Comment by Mehrad0711** [comment]
Hi @patil-suraj! 
Thanks for your review. As you suggested, I started updating the code and docs where `decode` or `batch_decode` is used.
Doing so, I noticed RAG model also has the same issue: in `decode` and `batch_decode`, `generator` is used instead of `current_tokenizer`. Do you want me to also update that model and its docs accordingly?

**Comment by Mehrad0711** [comment]
Hey, I submitted my changes and also fixed the RAG tokenizer. Please let me know if I missed something or you want me to change any of the fixes. I can rebase and force-push again.

**Comment by patrickvonplaten** [comment]
> do that inside the context manager, so we should also update all the ex

@patil-suraj @sgugger, wouldn't it be nicer to just do `as_target_tokenizer` in the `batch_decode` and `decode` function itself? Because decoding already corresponds to the "target tokenizer" -> I think this would be nicer for the user

**Comment by patil-suraj** [comment]
We could do that, but the reported issue is about not being able to decode the source tokens as the `decode` always uses `spm_target`.  So if we do `as_target_tokenizer` inside `decode` then source tokens will be decoded using `spm_target`, which will cause the same issue.

**Comment by patrickvonplaten** [comment]
True, yeah I was completely off there!

Ok, I understand the fix now. The problem now is that that's actually a pretty big backward breaking change, since in 99% of the cases people use `batch_decode` for the target outputs and we don't really want people to update the code just so that the source targets work correctly I think...

Could we maybe somehow, let the `current_spm` default to `target_spm` when using `batch_decode`, `decode` so that we don't have any breaking changes & then add maybe a new context manager `as_source_tokenizer`? or just add a optional arg to `decode` for Marian?

cc @LysandreJik @sgugger

**Comment by sgugger** [comment]
It's not very complicated to add the `as_source_tokenizer` context manager. Another solution is to add a flag `use_source_tokenizer` (defaults to False) to `decode` and `batch_decode`.

In any case, backward-compatibility is paramount so it needs to be fully enforced.

**Comment by patil-suraj** [comment]
`use_source_tokenizer` seems a better option to me, since the tokenizer already behaves like a source tokenizer by default so adding `as_source_tokenizer` seems a bit confusing IMO.

@Mehrad0711 , here's how we could now implement this
1. `current_spm` should always default to `source_spm` except inside the  `as_target_tokenizer`.
2. As Sylvain suggested, add the `use_source_tokenizer` argument to `decode` and `batch_decode`, if it's `True` use `source_spm` in `convert_tokens_to_string`. 
3. `convert_tokens_to_string` should never use `current_spm` as it defaults to `source_spm` and this would break backward-compatibility.

And the user should now pass `use_source_tokenizer` to decode source tokens

**Comment by Mehrad0711** [comment]
Thanks. I can proceed with the suggested implementation.
 However, passing `use_source_tokenizer` to `convert_tokens_to_string` requires updating `PreTrainedTokenizer`'s `_decode` method. Since `use_source_tokenizer` is passed to `convert_tokens_to_string` for all tokenizers, afaiu, the tokenizer classes for all models that implement their own `convert_tokens_to_string` should be updated individually to accept `use_source_tokenizer` (even though it's not used for some such as encoder-only models). 
I think a potential workaround can be adding instance checks within `_decode` to see if the tokenizer accepts that argument, or perhaps use function overloading.
Please let me know how you want me to proceed.

**Comment by sgugger** [comment]
I think we can work around this by setting an internal attribute with the value of `use_source_tokenizer` passed by the user. This way we can recover it in `convert_tokens_to_string` without having to overload any other methods. What do you think?

**Comment by Mehrad0711** [comment]
Thanks for the suggestion. It makes sense. If backward-compatibility wasn't a big issue, I think it would be better to use `current_spm` (set to None) in all tokenizer methods and switch it to source or target spm using two (`as_source_tokenizer` and `as_target_tokenizer`) context managers as needed. This way encoding and decoding methods become source/ target agnositc. 
However, since it's an issue, I think your first suggestion which is using `use_source_tokenizer` is still better than setting an internal value from the user perspective because now they have to use a context_manager during encoding but then set an attribute during decoding which can persist during next decoding if not unset. What do you think?

**Comment by sgugger** [comment]
> than setting an internal value from the user perspective

I never meant the user would have to set it. I meant for us to set it in `decode`/`decode_batch` with the `use_source_tokenizer` argument received.

**Comment by Mehrad0711** [comment]
Gotcha. Yeah, that should work.

**Comment by Mehrad0711** [comment]
Hi, the PR is ready for review. please let me know if the changes look good. Thanks.

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -3088,6 +3089,8 @@ def batch_decode(
                 Whether or not to remove special tokens in the decoding.
             clean_up_tokenization_spaces (:obj:`bool`, `optional`, defaults to :obj:`True`):
                 Whether or not to clean up the tokenization spaces.
+            use_source_tokenizer (:obj:`bool`, `optional`, defaults to :obj:`False`):
+                Whether or not to use source spm to decode sequences.
```
Comment (sgugger) [inline src/transformers/tokenization_utils_base.py:3093]:
```suggestion
                Whether or not to use the source tokenizer to decode sequences (only applicable in sequence-to-sequence problems).
```

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -3124,6 +3129,8 @@ def decode(
...
```
Comment (sgugger) [inline src/transformers/tokenization_utils_base.py:3133]:
```suggestion
                 Whether or not to use the source tokenizer to decode sequences (only applicable in sequence-to-sequence problems).
```

**Review by sgugger — COMMENTED** [review]
Thanks for amending the PR. Since this argument is very specific, I would:
- not put in Wav2Vec tokenizer
- pop it from the kwargs in the `_decode` methods instead of surfacing it and documenting it
- document it in the specific tokenizers where it's useful by creating a new docstring for the `decode` and `decode_batch` methods (copying the ones from the utils base and adding `use_source_tokenizer`.

I was also thinking the `decode_use_source_tokenizer` attribute should be private, but no strong opinion if people fell otherwise.

File: src/transformers/models/marian/tokenization_marian.py [hunk 3]
```diff
@@ -170,12 +170,19 @@ def _tokenize(self, text: str) -> List[str]:
...
+        """Converts an index (integer) in a token (str) using the decoder."""
         return self.decoder.get(index, self.unk_token)
 
     def convert_tokens_to_string(self, tokens: List[str]) -> str:
-        """Uses target language sentencepiece model"""
-        return self.spm_target.DecodePieces(tokens)
+        """
+        _decode_use_source_tokenizer is True if use_source_tokenizer is passed to decode/ batch_decode and is True.
+        use_source_tokenizer determines if the source tokenizer should be used to decode sequences (only applicable in
+        sequence-to-sequence problems).
```
Comment (patil-suraj) [inline src/transformers/models/marian/tokenization_marian.py:180]: I think we don't need to document it here, I think what Sylvain meant is, 

override the `decode` and `batch_decode` methods in the `MarianTokenizer` and document the `use_source_tokenizer` argument there. Users will never see the `convert_tokens_to_string` when decoding as it's mostly used internally.

But let's wait for @sgugger  to confirm :)

The rest of the changes look good to me. Thanks a lot!

Comment (sgugger) [reply src/transformers/models/marian/tokenization_marian.py:180]: Yes, this is indeed what I meant.

Comment (Mehrad0711) [reply src/transformers/models/marian/tokenization_marian.py:180]: No problem! I meant to send a comment after my changes and ask if you prefer these changes or you still want me to copy over decode and batch_decode to MarianTokenizer. I didn't take into account that users won't use convert_tokens_to_string directly and it's better not to hide the logic somewhere else as Patil suggested.
I will amend and push again.

**Comment by Mehrad0711** [comment]
Ok I think these changes address the comments. If there are still improvements to the docstring/ code you want to make, please feel free to make commit suggestions and I will amend. Thanks.

**Comment by sgugger** [comment]
The docstrings don't accumulate when you subclass and overload, they get reset. So we have to copy the whole things *and* add the extra argument. Will push on your branch the change.

**Comment by Mehrad0711** [comment]
Thanks. I think the PR is ready for the final review.
cc: @sgugger @patrickvonplaten @patil-suraj

**Review by LysandreJik — APPROVED** [review]
Cool! Thanks for taking care of it @Mehrad0711!

**Review by patil-suraj — APPROVED** [review]
Thanks for addressing all the review comments!

**Comment by sgugger** [comment]
Thanks a lot for your work on this PR!

**Comment by Mehrad0711** [comment]
Thanks a lot for your feedback and a great PR experience!

**Pull request merged**